### PR TITLE
HP9845C driver

### PR DESCRIPTION
**Pull request by fulivi**
Hi,

this hefty set of commits implements the HP-9845C driver.
This system was the color version of the 9845 range of machines.
The driver was written together with Ansgar Kueckes who did a lot of reverse engineering of the video and lightpen sub-systems. 

The best way to try the emulator is running the DEMO tape.
Quick instructions:
* Load the colorgfx optional ROM (-rom1 colorgfx)
* Load the demo1.hti into cassette drive (-magt demo1.hti)
* Optionally, enable lightpen emulation (-lightgun -lightgun_device mouse)
* Map the k8..k15 keys, they are not mapped by default and they are needed for menu navigation (see below).
* Start the demo with: LOAD "AUTOST",1 <EXECUTE>
* Be patient, demo takes a while to load..
* Once the system asks for the type of printer paper, load the demo2.hti image and answer the question (see next point)
* Use the 8 softkeys to navigate the menus. At the moment softkeys are not directly emulated but you can use the keyboard substitutes (SK1-8 map to SHIFT + K8-15 respectively)
* Enjoy!

I'm mailing all the necessary ROM & tape images to the usual "code" e-mail.
Beside this major change, this set of commits also fixes a subtle bug in nanoprocessor emulation.

Thanks.
--F.Ulivi

### CODE REVIEW

File: src/devices/bus/hp9845_io/hp9845_io.cpp [hunk 1]
```diff
@@ -74,7 +74,7 @@ void hp9845_io_card_device::flg_w(int state)
 void hp9845_io_card_device::install_readwrite_handler(read16_delegate rhandler, write16_delegate whandler)
 {
 	if (m_sys == nullptr) {
-		m_sys = dynamic_cast<hp9845b_state*>(&machine().root_device());
+		m_sys = dynamic_cast<hp9845_base_state*>(&machine().root_device());
```
Comment (cuavas) [inline src/devices/bus/hp9845_io/hp9845_io.cpp]: Making the assumption that the root device is the machine you want is pretty dangerous, especially as we're trying to move to the "everything is a device" model.  Are you sure you can't supply what you need through the slot interface somehow?

File: src/devices/bus/hp9845_io/hp9845_io.h [hunk 2]
```diff
@@ -48,7 +48,7 @@ class hp9845_io_slot_device : public device_t,
 	virtual void device_start() override;
 };
 
-class hp9845b_state;
+class hp9845_base_state;
```
Comment (cuavas) [inline src/devices/bus/hp9845_io/hp9845_io.h]: This is a dependency from src/devices on src/mame, isn't it?  We really need to eliminate these things.  Please try to find a way to access the necessary resources through your slot interface.

**Comment by fulivi** [comment]
Ok, I'll think about this. Hopefully it shouldn't be too difficult to re-work so that dependencies are the other way around. There maybe some problem at device start because of the order the devices are started wrt the base system..

**Comment by startaq** [comment]
If you depend on another device being started first, you can throw an exeption of type `device_missing_dependencies()` in your `device_start()` to reschedule your device to be started later. You can test if a device has been started with `device->started()`.

**Comment by fulivi** [comment]
Thanks for the suggestion! Looks really helpful in my case..

**Comment by fulivi** [comment]
Is it better like this? I eliminated the dependency on src/mame from src/devices. As far as I can tell the driver behaves as before.

**Comment by cuavas** [comment]
That's great, thanks.

**Pull request merged**